### DISCUSSION TITLE
The widget only permits use on 127.0.0.1 and localhost.  In case this doesn't work for you, let us know on GitHub.

### DISCUSSION THREAD

**Discussion by NanisTe** [yWorks/yfiles-jupyter-graphs#28] [Q&A] [answered]
We tried to use the widget in our online Kubeflow environment on our AWS Kubernetes Cluster. Unfortunately I got this message on running the widgets show function:

The widget only permits use on 127.0.0.1 and localhost.
In case this doesn't work for you, let us know on [GitHub](https://github.com/yWorks/yfiles-jupyter-graphs/issues).

Could someone please explain the boundary conditions to run the widget?

Thanks
Best Sinan

**Comment by yGuy** [comment]
Well, as the error message states, the widget won't run when the page is served from a server that is different from 127.0.01 or localhost. This is obviously the case for your setup. 

You may ask why? 

This is because we want to prevent commercial usage where someone else is hosting the server/widget for you (and might even be charging you for using our free widget). This was a commercial, rather than a technical decision. 

We do want to support common use-cases. The most common use-case obviously is to run the server on your own computer. We also recently allowed the usage of the widget in `mybinder.org` and once we have support for it implemented, we also want to support kaggle and google colab.

If you want us to consider your use-case, we would need more information. If you do not want to share this publicy, feel free to contact us via our contact form on https://www.yworks.com

Thanks.

**Comment by yGuy** [comment]
We are currently considering lifting the restriction and are looking for white listing of public jupyter notebook servers like kaggle, mybinder.org, etc.

If you know of a big domain that hosts jupyter notebook services and where the plugin doesn't run, feel free to drop a note, here, and we'll likely add it to our white list with one of the next releases.

**Comment by yGuy** [comment]
With the 1.4.2 release we enabled Visual Studio Code and google colab, as well as dominolab. myBinder should also still work. 
Also HTML exports of the notebooks should now work, standalone.
Kaggle could work, however it seems Kaggle itself is broken and we were not able to get any custom widget to work, there.
GitHub unfortunately does not support custom widgets and we weren't able to get the widgets to work in the JetBrains products, either.

**Comment by TermeHansen** [comment]
I don't think I get this restriction. I my use case, I use Jupyterlab on a local calculation server within our company LAN. Why should we have this restriction from using this widget on our own servers or when working remote? This is the true beauty of jupyterlab for me, that I can work on the same instance of jupyterlab as long as I have a browser with access to my server...

What are my options in this use case?

**Reply by yGuy** [reply]
> This is because we want to prevent commercial usage where someone else is hosting the server/widget for you (and might even be charging you for using our free widget). This was a commercial, rather than a technical decision.

I can see that you have installed the widget yourself in this case, but you could just as well share your notebook URL with someone else so that they can use your notebook with our widget like a SAAS solution; they might even charge you for that. We want to avoid this scenario. We want our users to be aware of the fact that they are using our (free as in beer) product. If you have an idea how we can avoid this scenario and detect that you as the user actually installed the widget yourself for free (and hence accepted the license terms), we are all open to this. 

It seems you are a power-user with a custom jupyter server infrastructure. If you can help us detect this situation (mostly by looking at the URL that is shown in the browser), we will see what we can do. 

Right now, it sounds as if this was a use-case that we would not want to restrict, so if this is a common use-case, we would need a way to detect it to "white-list" this.

Again, this is mostly a commercial/political decision which itself I do not want to discuss, here, and unfortunately you seem to run into the situation where our heuristic fails to detect your use-case. We would be happy to address the failure in the detection to make a use-case like this possible.

**Reply by yGuy** [reply]
> What are my options in this use case?

If you need a workaround, right now, it might work to proxy your jupyter server to localhost - install a proxy on your machine and connect to your server via a proxy on some localhost port that forwards all the requests to the non-localhost server.

Or host your notebook on google colab, or mybinder, which are white-listed, too.

**Comment by rfrneo4j** [comment]
I currently seem to be able to use this library via Google Colab notebook but can't use it in an Azure Machine Learning Studio Notebook (which feels weird that it works in one and not the other). I love this library and am disappointed it didn't work in the latter.

**Reply by yGuy** [reply]
What does the URL you see in the browser look like when you are working with your notebook? 

Like so: instances.azureml.net and *.instances.azureml.ms ?

**Reply by rfrneo4j** [reply]
https://ml.azure.com/fileexplorerAzNB?

there are a bunch of other query parameters at the end of the url when the notebook is open in my browser

**Reply by fskpf** [reply]
Thanks for the details. Sounds like we could easily whitelist this in the upcoming days.

Just to be sure: The invalid host popup in the widget should tell you which location is actually blocked. It states 
```
"You are running yFiles Graphs for Jupyter on [location] ..."
```
Can you send us the mentioned location? Then we know exactly what URL / domain we need to whitelist.

**Reply by rfrneo4j** [reply]
@fskpf my bad. 

You are running yFiles Graphs for Jupyter on https://{id}.eastus.notebooks.azuresandbox.ms.

**Reply by fskpf** [reply]
We just deployed a new version ([v1.4.4](https://pypi.org/project/yfiles-jupyter-graphs/1.4.4/)) that should work on the mentioned domain.

Please try it again with the update. Let us know if it works for you since we cannot easily test it.

**Reply by yGuy** [reply]
@runfourestrun  - did you find the time to check whether the white-listing worked? Thanks!

**Reply by rfrneo4j** [reply]
@yGuy It did! Thanks for your responsiveness.

**Comment by yGuy** [comment]
I was just trying to use the widget on Amazon AWS Sagemaker. It's not yet white-listed.
My URL looked like this: 
https://[somemagicnumbershere].studio.us-east-2.sagemaker.aws

I guess whitelisting `sagemaker.aws` should do it? (us-east-2 might be specific, although I am definitely close to that region). "studio" could also be checked...

**Reply by fskpf** [reply]
Whitelisted with [v1.4.5](https://pypi.org/project/yfiles-jupyter-graphs/1.4.5/).

**Comment by yGuy** [comment]
John Hopkins university has a hosted Jupyter Notebook installation at https://apps.sciserver.org/ - we should whitelist that one.

> "You are running yFiles Graphs for Jupyter on https://apps.sciserver.org/."

**Reply by fskpf** [reply]
Allowed with [v1.4.6](https://pypi.org/project/yfiles-jupyter-graphs/1.4.6/).

**Comment by alexdashkov** [comment]
Hello, 
Is it possible to whitelist the plugin for the Google's Vertex AI notebooks? 
The URL looks like that `https://[somemagicnumbersandletters].notebooks.googleusercontent.com`

**Reply by yGuy** [reply]
Thanks for the issue. I agree - we will whitelist this. Please stay tuned for the next release!

**Reply by fskpf** [reply]
We allowed the mentioned URL pattern with [v1.4.6](https://pypi.org/project/yfiles-jupyter-graphs/1.4.6/). 
If it still does not work, please provide us the "You are running yFiles Graphs for Jupyter on `URL`" message from the error popup in the widget.

**Reply by alexdashkov** [reply]
I've just tested with the new version and I can confirm that it works. Thanks!

**Comment by engg14000** [comment]
Hello @yGuy  ,
Please whitelist the plugin for the Google's DataProc Jupyter notebooks?
URL format : https://<cluster_id>.dataproc.googleusercontent.com

**Reply by fskpf** [reply]
Allowed with [v1.4.7](https://pypi.org/project/yfiles-jupyter-graphs/1.4.7/).

**Reply by engg14000** [reply]
from yfiles_jupyter_graphs import GraphWidget
#shows empty widget
GraphWidget()

This yields below error
Error displaying widget : model not found

Any idea how to resolve this?

**Reply by fskpf** [reply]
Please try to restart Jupyter Lab (mandatory after installing a new widget). If this does not help, please open a new issue [here](https://github.com/yWorks/yfiles-jupyter-graphs/issues) where we can dig deeper.

**Reply by yGuy** [reply]
The lab and/or the kernel. The new widget and its dependencies need to be registered properly in the kernel and lab and depending on which ones where installed during the previous start, a restart might be required, unfortunately. I don't think we can improve the situation here, other than maybe improve the documentation. We have this mentioned in some places, but obviously you don't get to see this at runtime. Jupyter Lab's output could be more helpful, though.

**Reply by engg14000** [reply]
I did cluster restart and it helped. Thanks 👍

**Comment by craig-willis** [comment]
I work on a number of public research data infrastructure projects based on the [Whole Tale](https://wholetale.org/) platform where we would like to include support for the yfiles jupyter widget. Would it be possible to whitelist the following domains?
* wholetale.org
* htmdec.org

**Reply by yGuy** [reply]
I think we will whitelist wholetale.org. Thanks for suggesting it!

The "www.htmdec.org" looks pretty closed, though. Can you elaborate on how it is used and by whom? Feel free to contact us via our contact forms at yworks.com, if you don't want to disclose this information, here.

**Reply by craig-willis** [reply]
Thank you, @yGuy.  Sorry - there is nothing at www.htmdec.org right now (we should correct that). You can find more information here:

* https://docs.htmdec.org/
* https://github.com/htmdec/

This is a deployment of the Whole Tale software stack for materials science research. The data is access controlled, but the software is public and funded by research grants.

**Comment by DylanShang** [comment]
Hello @yGuy 
Please whitelist [kuzu-lab.netlify.app](https://kuzu-lab.netlify.app/)

[Kuzu-wasm](https://github.com/DylanShang/kuzu-wasm) is the WebAssembly version of Kuzu and is also the first WebAssembly database in the graph domain. With the help of JupyterLite, I created [Kuzu-lab](https://github.com/DylanShang/kuzu-lab/), enabling data science research with zero installation costs. All of the above projects are open source and are used for analysis in the graph data domain, with researchers from the [University of New South Wales](https://unswdb.github.io/). By using yFiles Jupyter, we can enhance user interactivity and increase the visibility of yFiles Jupyter to a broader audience.

**Reply by fskpf** [reply]
I'll whitelist the URL with the next feature release (`v1.8.0`) which is planned for the end of this (or maybe the upcoming) week.

**Reply by DylanShang** [reply]
Thanks very much!

**Comment by yGuy** [comment, marked as answer]
The restriction has been lifted in recent versions. You should be able to run the widget on any domain *for testing purposes* on any non-white-listed domain, by obtaining a temporary license inside the widget (just click the button that is being shown in the case you are using it on a non-white-listed server).

If you want to host the widget for commercial purposes or in a commercial context on an intranet server to a larger audience, please turn to the [team at yworks.com](https://yworks.com/contact) and they can provide you with a commercial license for deploying the widget to private users.

**Comment by JulienHomo** [comment]
Hello yWorks team,

We are reaching out on behalf of the [FASCA project](https://gotriple.eu/data-services/fascaInitiative) (Facilitating Scholarly Communication Analysis through GoTriple Pipeline),  funded by the European Union and coordinated by [OPERAS AISBL](https://operas-eu.org/) – the European research infrastructure for open scholarly communication in the Social Sciences and Humanities (SSH).

As part of the project, we are deploying a dedicated JupyterHub instance for data-driven SSH research support:
https://fasca-ds-tools.operas-eu.org/

Could you please whitelist the domain fasca-ds-tools.operas-eu.org for the yfiles-jupyter-graphs plugin?
It’s used in a research-only, non-commercial JupyterHub instance as part of the FASCA project.

Let us know if you need any additional details.

Thanks a lot for your support!

**Reply by yGuy** [reply]
Thanks! I'll guess we'll do that. We'll get back to you in the email thread you started.

**Comment by fschuh** [comment]
hi @yGuy,

I was trying to use yfiles for Jupyter for documentation purposes in our company's intranet, in JupyterLite which is hosted in Gitlab pages (hosted in our intranet), and ran into this error.
Is this use case allowed with the free license?
Would we need to whitelist our company's intranet domain somehow?

Thanks in advance!

**Reply by yGuy** [reply]
Hi @fschuh - 

Actually, you should not be seeing an error, anymore. If you see an error, this might indicate a bug that we would like to fix.

Can you please provide more details? As the marked answer for this thread explains, you should now always be able to get at least a temporary license if you are using the widget in a local closed intranet. 

We are happy to whitelist public domains with a focus on non-commercial use. The *public* Jupyter pages or GitLab pages would fall under this category. If you are looking to get a license for your non-publicly hosted jupyter notebooks and want to use more than what the free temporary licenses offer, please contact [yworks](https://yworks.com/contact), directly. At this time, we are not whitelisting commercial non-public offerings of the widget beyond the free testing licenses.

Thanks!

**Reply by fschuh** [reply]
Sorry for the confusion, I'm actually able to select the free temporary license, and it works. However, I believe the error message is a bit misleading.

Everybody I show this to thinks we're not allowed to use it because we're not running on 127.0.0.1 or localhost, and the free temporary license seems to imply it's a workaround that we **should not be using**, as if we're not being compliant to the free license.

If that's not the case, I believe it should be made more clear in that error message.

**Reply by yGuy** [reply]
Thanks for the feedback. I understand the confusion. But, yes, it is a free *temporary* license that is meant to be used to evaluate the software in closed environments. If you want to get a permanent license that lets you look less like you are using the software in an illegitimate way, please consider getting a license for commercial in-house use. Or just use it on localhost or on google colab, or any other of the white-listed providers if this feels not worth the investment :-)
